tendrl/ceph_integration/sds_sync/__init__.py

```diff
@@ -239,6 +245,230 @@ def _sync_ec_profiles(self):
+            "clusters/%s/maps/osd_map/data/osds" %
+            NS.tendrl_context.integration_id
+        ).value.
+        encode('utf-8').
+        replace("'", '"').
+        replace('True', 'true').
+        replace('False', 'false'))
+        old_osd_ids = set([o['osd'] for o in old_osds_det])
+        new_osd_ids = set([o['osd'] for o in data['osds']])
+        deleted_osds = old_osd_ids - new_osd_ids
+        created_osds = new_osd_ids - old_osd_ids
+
+        def osd_event(severity, msg, osd_id, curr_value):
+            self._emit_event(
+                severity,
+                'osd_status',
+                curr_value,
+                msg.format(
+                    name=NS.tendrl_context.cluster_name,
+                    id=osd_id
+                )
+            )
+
+        # Generate events for removed OSDs
+        for osd_id in deleted_osds:
+            osd_event(
+                INFO,
+                "OSD {name}.{id} removed from the cluster map",
+                osd_id,
+                ""
+            )
+
+        # Generate events for added OSDs
+        for osd_id in created_osds:
+            osd_event(
+                INFO,
+                "OSD {name}.{id} added to the cluster map",
+                osd_id,
+                ""
+            )
+
+        # Generate events for changed OSDs
+        for osd_id in old_osd_ids & new_osd_ids:
+            old_osd = next(
+                (osd for osd in old_osds_det if osd['osd'] == osd_id),
+                None
+            )
+            new_osd = next(
+                (osd for osd in data['osds'] if osd['osd'] == osd_id),
+                None
+            )
+            if old_osd['up'] != new_osd['up']:
+                if bool(new_osd['up']):
+                    osd_event(
+                        INFO,
+                        "OSD {name}.{id} came up",
+                        osd_id,
+                        str(new_osd['up'])
+                    )
+                else:
+                    osd_event(
+                        WARNING,
+                        "OSD {name}.{id} went down",
+                        osd_id,
+                        str(new_osd['up'])
+                    )
+            if old_osd['in'] != new_osd['in']:
+                if bool(new_osd['in']):
+                    osd_event(
+                        INFO,
+                        "OSD {name}.{id} is in",
+                        osd_id,
+                        str(new_osd['in'])
+                    )
+                else:
+                    osd_event(
+                        WARNING,
+                        "OSD {name}.{id} went out",
+                        osd_id,
+                        str(new_osd['in'])
+                    )
+
+    def _on_pool_status(self, data):
+        old_pools_det = json.loads(NS._int.client.read(
+            "clusters/%s/maps/osd_map/data/pools" %
+            NS.tendrl_context.integration_id
+        ).value.
```

Comment:
I tried using `old_osds_det = NS.ceph.objects.SyncObject(sync_type='osd_map').load().data['osds']` but value of `NS.ceph.objects.SyncObject(sync_type='osd_map').load().data` is `None`. Not sure why it not set after load.